Nama  : Reza Hilmi Dafa

NIM   : 24060121130044

Nama Dataset : iris

Link Dataset : https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data

Baris ini membaca dataset Iris dari URL

In [ ]:
import pandas

In [ ]:
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data"
names = ['sepal-length', 'sepal-width', 'petal length', 'petal-width', 'class']
dataset = pandas.read_csv(url, names=names)

In [ ]:
dataset.head(5)

sepal-length  sepal-width  petal length  petal-width        class
0           5.1          3.5           1.4          0.2  Iris-setosa
1           4.9          3.0           1.4          0.2  Iris-setosa
2           4.7          3.2           1.3          0.2  Iris-setosa
3           4.6          3.1           1.5          0.2  Iris-setosa
4           5.0          3.6           1.4          0.2  Iris-setosa

In [ ]:
dataset.shape

(150, 5)

Baris di bawah ini mengimpor modul model_selection dari scikit-learn

In [ ]:
from sklearn import model_selection

-Menggunakan atribut values dari objek DataFrame Pandas (dataset), kita mengonversi dataset ke dalam bentuk array NumPy

-Memisahkan dataset menjadi dua bagian:

X: Variabel independen (fitur) yang terdiri dari kolom 0 hingga 3.

Y: Variabel dependen (kelas/target) yang terdiri dari kolom ke-4.

-kemudian menggunakan fungsi train_test_split dari scikit-learn, dataset (X dan Y) dibagi menjadi dua bagian:

X_train: Data pelatihan untuk variabel independen.

X_validation: Data validasi untuk variabel independen.

Y_train: Data pelatihan untuk variabel dependen.

Y_validation: Data validasi untuk variabel dependen.

test_size=validation_size menentukan proporsi data yang akan dialokasikan sebagai data validasi (20% dalam hal ini).

random_state=seed memastikan reproduktibilitas pembagian data dengan memberikan nilai seed pada generator angka acak.

In [ ]:
array = dataset.values

X = array[:,0:4]
Y = array[:,4]

validation_size = 0.20

seed = 7

X_train, X_validation, Y_train, Y_validation = model_selection.train_test_split(X, Y, test_size=validation_size,
random_state=seed)

Kemudian hasilnya kita tampilkan

In [ ]:
print("X_train :", X_train.shape)
print("X_validation :", X_validation.shape)
print("Y_train :", Y_train.shape)
print("Y_validation :", Y_validation.shape)

X_train : (120, 4)
X_validation : (30, 4)
Y_train : (120,)
Y_validation : (30,)


Menetapkan seed yaitu 7 pada algoritma untuk memastikan bahwa hasilnya dapat direproduksi.

Menunjukkan bahwa metrik evaluasi yang digunakan untuk mengukur performa model adalah akurasi.

In [ ]:
# Test options and evaluation metric
seed = 7
scoring = 'accuracy'

mengimpor kelas DecisionTreeClassifier dari modul tree dalam scikit-learn.

kemudian membuat sebuah list models yang akan berisi tuple-tuple model yang akan dievaluasi. Dalam kasus ini, kita memiliki satu model, yaitu Decision Tree.

In [ ]:
from sklearn.tree import DecisionTreeClassifier

models = []

models.append(('Decision Tree', DecisionTreeClassifier()))


* List results akan digunakan untuk menyimpan hasil evaluasi model.

* List names akan digunakan untuk menyimpan nama-nama model.

* Melakukan iterasi melalui setiap model dalam list models.

* Membuat objek KFold untuk validasi silang dengan 10 lipatan (folds), menggunakan nilai seed yang telah ditentukan (random_state=seed) dan dengan pengacakan data (shuffle=True).

* Menggunakan cross_val_score untuk mengevaluasi model menggunakan validasi silang. Hasil evaluasi disimpan dalam cv_results.

* Menambahkan hasil evaluasi dan nama model ke dalam list results dan names.
Mencetak hasil evaluasi model, termasuk rata-rata (mean()) dan deviasi standar (std()).




In [ ]:
# evaluate each model in turn
results = []

names = []
for name, model in models:

  kfold = model_selection.KFold(n_splits=10, random_state=seed, shuffle=True)

  cv_results = model_selection.cross_val_score(model, X_train, Y_train, cv=kfold, scoring=scoring)

  results.append(cv_results)

  names.append(name)

  msg = "%s: %f (%f)" % (name, cv_results.mean(), cv_results.std())
  print(msg)

Decision Tree: 0.958333 (0.076830)


mengimpor beberapa metrik evaluasi dari scikit-learn yang akan digunakan untuk mengevaluasi performa model.

In [ ]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

* Membuat objek model Decision Tree (decision_tree) dari kelas DecisionTreeClassifier.
* Melatih model menggunakan data pelatihan (X_train dan Y_train) menggunakan metode fit()
* Menggunakan model yang telah dilatih untuk membuat prediksi pada data validasi (X_validation).
* Mencetak akurasi model dengan membandingkan prediksi dengan nilai sebenarnya menggunakan accuracy_score.
* Mencetak matriks kebingungan (confusion matrix) menggunakan confusion_matrix, yang memberikan gambaran tentang seberapa baik model mengklasifikasikan kelas-kelas.
* Mencetak laporan klasifikasi (classification_report), yang menyediakan informasi seperti presisi, recall, dan f1-score untuk setiap kelas.

In [ ]:
decision_tree = DecisionTreeClassifier()
decision_tree.fit(X_train, Y_train)
predictions = decision_tree.predict(X_validation)

print(accuracy_score(Y_validation, predictions))
print(confusion_matrix(Y_validation, predictions))
print(classification_report(Y_validation, predictions))

0.8666666666666667
[[ 7  0  0]
 [ 0 10  2]
 [ 0  2  9]]
                 precision    recall  f1-score   support

    Iris-setosa       1.00      1.00      1.00         7
Iris-versicolor       0.83      0.83      0.83        12
 Iris-virginica       0.82      0.82      0.82        11

       accuracy                           0.87        30
      macro avg       0.88      0.88      0.88        30
   weighted avg       0.87      0.87      0.87        30



penjelasan informasi atau kesimpulan dari data yang ditampilkan
* **Model Decision Tree mencapai akurasi sekitar 95.83% pada data pelatihan menggunakan validasi silang dengan 10 lipatan.**
* **Dalam evaluasi pada data validasi, akurasi model adalah sekitar 86.67%, yang masih baik.**